src/main/java/org/apache/log4j/jdbc/JDBCAppender.java

```diff
@@ -274,27 +283,76 @@ public void close() {
      * If a statement fails the LoggingEvent stays in the buffer!
      */
     public void flushBuffer() {
+		if (buffer.isEmpty()) {
+			return;
+		}
 	// Do the actual logging
 	removes.ensureCapacity(buffer.size());
-	for (Iterator i = buffer.iterator(); i.hasNext();) {
-	    LoggingEvent logEvent = (LoggingEvent) i.next();
-	    try {
-		String sql = getLogStatement(logEvent);
-		execute(sql);
-	    } catch (SQLException e) {
-		errorHandler.error("Failed to excute sql", e, ErrorCode.FLUSH_FAILURE);
-	    } finally {
-		removes.add(logEvent);
-	    }
-	}
-
+		if (SECURE_SQL_REPLACEMENT) {
+			flushBufferSecure();
+		} else {
+			flushBufferInsecure();
+		}
 	// remove from the buffer any events that were reported
 	buffer.removeAll(removes);
 
 	// clear the buffer of reported events
 	removes.clear();
     }
 
+	private void flushBufferInsecure() {
+		for (Iterator i = buffer.iterator(); i.hasNext(); ) {
+			LoggingEvent logEvent = (LoggingEvent) i.next();
+			try {
+				String sql = getLogStatement(logEvent);
+				execute(sql);
+			} catch (SQLException e) {
+				errorHandler.error("Failed to excute sql", e, ErrorCode.FLUSH_FAILURE);
+			} finally {
+				removes.add(logEvent);
+			}
+		}
+	}
+
+	private void flushBufferSecure() {
+		if (layout.getClass() != PatternLayout.class) {
+			errorHandler.error("Failed to convert pattern " + layout + " to SQL", ILLEGAL_PATTERN_FOR_SECURE_EXEC, ErrorCode.MISSING_LAYOUT);
+			return;
+		}
+		PatternLayout patternLayout = (PatternLayout) layout;
+		preparedStatementParser.setPattern(patternLayout.getConversionPattern());
+		Connection con = null;
+		try {
+			con = getConnection();
+			PreparedStatement ps = con.prepareStatement(preparedStatementParser.getParameterizedSql());
+			try {
+				for (Iterator i = buffer.iterator(); i.hasNext(); ) {
+					LoggingEvent logEvent = (LoggingEvent) i.next();
+					try {
+						preparedStatementParser.setParameters(ps, logEvent);
+						ps.addBatch();
+					} catch (SQLException e) {
+						errorHandler.error("Failed to append parameters", e, ErrorCode.FLUSH_FAILURE);
+					} finally {
+						removes.add(logEvent);
```

Comment:
This should probably be moved out if loop as removes.addAll(buffer) or something like that

Reply:
I will be spending tomorrow morning looking at this PR.